doc/api/async_hooks.md

```diff
@@ -459,6 +459,62 @@ init for PROMISE with id 6, trigger id: 5  # the Promise returned by then()
   after 6
 ```
 
+#### `async_hooks.executionAsyncResource()`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+* Returns: {Object} The resource representing the current execution.
+  Useful to store data within the resource.
+
+```js
+const { open } = require('fs');
+const { executionAsyncId, executionAsyncResource } = require('async_hooks');
+
+console.log(executionAsyncId(), executionAsyncResource());  // 1 {}
+open(__filename, 'r', (err, fd) => {
+  console.log(executionAsyncId(), executionAsyncResource());  // 7 FSReqWrap
+});
+```
+
+This can be used to implement continuation local storage without the
+using of a tracking `Map` to store the metadata:
+
+```js
+const { createServer } = require('http');
+const {
+  executionAsyncId,
+  executionAsyncResource,
+  createHook
+} = require('async_hooks');
+const sym = Symbol('state'); // Private symbol to avoid pollution
+
+createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    const cr = executionAsyncResource();
+    if (cr) {
+      resource[sym] = cr[sym];
+    }
+  }
+}).enable();
+
+const server = createServer(function(req, res) {
+  executionAsyncResource()[sym] = { state: req.url };
+  setTimeout(function() {
+    res.end(JSON.stringify(executionAsyncResource()[sym]));
+  }, 100);
+}).listen(3000);
+```
+
+Resource objects returned by `executionAsyncResource()` are often internal
+handle objects with undocumented APIs. Using any functions or properties
+on the object is not recommended and may crash your application.
```

Comment:
This likely needs to be *much* stronger here. I would move this up closer to the top and make a stronger statement:

```
Resource objects returned by `executionAsyncResource()` are most often internal
Node.js handle objects with undocumented APIs. Using any functions or properties
on the object is likely to crash your application and should be avoided.
```

The other question here is whether we should expose the `owner_symbol` to allow user code to get access to the public object if one exists? e.g. 

```js
const owner_symbol = ...
const obj = executionAsyncResource()[owner_symbol];
```

Reply:
Updated. Didn't do anything about `owner_symbol`--I think we should consider that separately and not let it block this PR.

Reply:
Yeah, we should not expose `owner_symbol`, and it’s our of the scope for this PR, too.